File: ios/Classes/HeadlessInAppWebView/HeadlessInAppWebView.swift
//
//  HeadlessInAppWebView.swift
//  flutter_inappwebview
//
//  Created by Lorenzo Pichilli on 26/03/21.
//

import Foundation

public class HeadlessInAppWebView : FlutterMethodCallDelegate {
    var id: String
    var channel: FlutterMethodChannel?
    var flutterWebView: FlutterWebViewController?
    
    public init(id: String, flutterWebView: FlutterWebViewController) {
        self.id = id
        super.init()
        self.flutterWebView = flutterWebView
        self.channel = FlutterMethodChannel(name: "com.pichillilorenzo/flutter_headless_inappwebview_" + id,
                                       binaryMessenger: SwiftFlutterPlugin.instance!.registrar!.messenger())
        self.channel?.setMethodCallHandler(self.handle)
    }
    
    public override func handle(_ call: FlutterMethodCall, result: @escaping FlutterResult) {
        let arguments = call.arguments as? NSDictionary
        
        switch call.method {
        case "dispose":
            dispose()
            result(true)
            break
        case "setSize":
            let sizeMap = arguments!["size"] as? [String: Any?]
            if let size = Size2D.fromMap(map: sizeMap) {
                setSize(size: size)
            }
            result(true)
            break
        case "getSize":
            result(getSize()?.toMap())
            break
        default:
            result(FlutterMethodNotImplemented)
            break
        }
    }
    
    public func onWebViewCreated() {
        let arguments: [String: Any?] = [:]
        channel?.invokeMethod("onWebViewCreated", arguments: arguments)
    }
    
    public func prepare(params: NSDictionary) {
        if let view = flutterWebView?.view() {
            view.alpha = 0.01
            let initialSize = params["initialSize"] as? [String: Any?]
            if let size = Size2D.fromMap(map: initialSize) {
                setSize(size: size)
            } else {
                view.frame = CGRect(x: 0.0, y: 0.0, width: UIScreen.main.bounds.width, height: UIScreen.main.bounds.height)
            }
            if let keyWindow = UIApplication.shared.keyWindow {
                /// Note: The WKWebView behaves very unreliable when rendering offscreen
                /// on a device. This is especially true with JavaScript, which simply
                /// won't be executed sometimes.
                /// So, add the headless WKWebView to the view hierarchy.
                /// This way is also possible to take screenshots.
                keyWindow.insertSubview(view, at: 0)
                keyWindow.sendSubviewToBack(view)
            }
        }
    }
    
    public func setSize(size: Size2D) {
        if let view = flutterWebView?.view() {
            let width = size.width == -1.0 ? UIScreen.main.bounds.width : CGFloat(size.width)
            let height = size.height == -1.0 ? UIScreen.main.bounds.height : CGFloat(size.height)
            view.frame = CGRect(x: 0.0, y: 0.0, width: width, height: height)
        }
    }
    
    public func getSize() -> Size2D? {
        if let view = flutterWebView?.view() {
            return Size2D(width: Double(view.frame.width), height: Double(view.frame.height))
        }
        return nil
    }
    
    public func dispose() {
        channel?.setMethodCallHandler(nil)
        channel = nil
        HeadlessInAppWebViewManager.webViews[id] = nil
        flutterWebView = nil
    }
    
    deinit {
        print("HeadlessInAppWebView - dealloc")
        dispose()
    }
}

File: android/src/main/java/com/pichillilorenzo/flutter_inappwebview/headless_in_app_webview/HeadlessInAppWebView.java
package com.pichillilorenzo.flutter_inappwebview.headless_in_app_webview;

import android.view.View;
import android.view.ViewGroup;
import android.widget.FrameLayout;

import androidx.annotation.NonNull;
import androidx.annotation.Nullable;

import com.pichillilorenzo.flutter_inappwebview.InAppWebViewFlutterPlugin;
import com.pichillilorenzo.flutter_inappwebview.Util;
import com.pichillilorenzo.flutter_inappwebview.in_app_webview.FlutterWebView;
import com.pichillilorenzo.flutter_inappwebview.types.Size2D;

import java.util.HashMap;
import java.util.Map;

import io.flutter.plugin.common.MethodCall;
import io.flutter.plugin.common.MethodChannel;

public class HeadlessInAppWebView implements MethodChannel.MethodCallHandler {

  protected static final String LOG_TAG = "HeadlessInAppWebView";
  @NonNull
  public final String id;
  public final MethodChannel channel;
  @Nullable
  public FlutterWebView flutterWebView;
  @Nullable
  public InAppWebViewFlutterPlugin plugin;

  public HeadlessInAppWebView(@NonNull final InAppWebViewFlutterPlugin plugin, @NonNull String id, @NonNull FlutterWebView flutterWebView) {
    this.id = id;
    this.plugin = plugin;
    this.flutterWebView = flutterWebView;
    this.channel = new MethodChannel(plugin.messenger, "com.pichillilorenzo/flutter_headless_inappwebview_" + id);
    channel.setMethodCallHandler(this);
  }

  @Override
  public void onMethodCall(@NonNull MethodCall call, @NonNull MethodChannel.Result result) {
    switch (call.method) {
      case "dispose":
        dispose();
        result.success(true);
        break;
      case "setSize":
        {
          Map<String, Object> sizeMap = (Map<String, Object>) call.argument("size");
          Size2D size = Size2D.fromMap(sizeMap);
          if (size != null)
            setSize(size);
        }
        result.success(true);
        break;
      case "getSize":
        {
          Size2D size = getSize();
          result.success(size != null ? size.toMap() : null);
        }
        break;
      default:
        result.notImplemented();
    }
  }

  public void onWebViewCreated() {
    Map<String, Object> obj = new HashMap<>();
    channel.invokeMethod("onWebViewCreated", obj);
  }

  public void prepare(Map<String, Object> params) {
    if (flutterWebView != null) {
      View view = flutterWebView.getView();
      if (view != null) {
        final Map<String, Object> initialSize = (Map<String, Object>) params.get("initialSize");
        Size2D size = Size2D.fromMap(initialSize);
        if (size == null) {
          size = new Size2D(-1, -1);
        }
        setSize(size);
        view.setVisibility(View.INVISIBLE);
      }
    }
    if (plugin != null && plugin.activity != null) {
      // Add the headless WebView to the view hierarchy.
      // This way is also possible to take screenshots.
      ViewGroup contentView = (ViewGroup) plugin.activity.findViewById(android.R.id.content);
      if (contentView != null) {
        ViewGroup mainView = (ViewGroup) (contentView).getChildAt(0);
        if (mainView != null && flutterWebView != null) {
          View view = flutterWebView.getView();
          if (view != null) {
            mainView.addView(view, 0);
          }
        }
      }
    }
  }

  public void setSize(@NonNull Size2D size) {
    if (flutterWebView != null && flutterWebView.webView != null) {
      View view = flutterWebView.getView();
      if (view != null) {
        float scale = Util.getPixelDensity(view.getContext());
        Size2D fullscreenSize = Util.getFullscreenSize(view.getContext());
        int width = (int) (size.getWidth() == -1 ? fullscreenSize.getWidth() : (size.getWidth() * scale));
        int height = (int) (size.getWidth() == -1 ? fullscreenSize.getHeight() : (size.getHeight() * scale));
        view.setLayoutParams(new FrameLayout.LayoutParams(width, height));
      }
    }
  }

  @Nullable
  public Size2D getSize() {
    if (flutterWebView != null && flutterWebView.webView != null) {
      View view = flutterWebView.getView();
      if (view != null) {
        float scale = Util.getPixelDensity(view.getContext());
        Size2D fullscreenSize = Util.getFullscreenSize(view.getContext());
        ViewGroup.LayoutParams layoutParams = view.getLayoutParams();
        return new Size2D(
                fullscreenSize.getWidth() == layoutParams.width ? layoutParams.width : (layoutParams.width / scale),
                fullscreenSize.getHeight() == layoutParams.height ? layoutParams.height : (layoutParams.height / scale)
        );
      }
    }
    return null;
  }

  public void dispose() {
    channel.setMethodCallHandler(null);
    if (HeadlessInAppWebViewManager.webViews.containsKey(id)) {
      HeadlessInAppWebViewManager.webViews.put(id, null);
    }
    if (plugin != null && plugin.activity != null) {
      ViewGroup contentView = (ViewGroup) plugin.activity.findViewById(android.R.id.content);
      if (contentView != null) {
        ViewGroup mainView = (ViewGroup) (contentView).getChildAt(0);
        if (mainView != null && flutterWebView != null) {
          mainView.removeView(flutterWebView.getView());
        }
      }
    }
    if (flutterWebView != null) {
      flutterWebView.dispose();
    }
    flutterWebView = null;
    plugin = null;
  }
}
